lambda_function.py

```diff
@@ -0,0 +1,135 @@
+        else:
+            regular_open_est = _EASTERN_TIMEZONE.localize(
+                datetime.combine(polling_date, last_alert['regular_open']))
+            regular_close_est = _EASTERN_TIMEZONE.localize(
+                datetime.combine(polling_date, last_alert['regular_close']))
+            alert_start_est = last_alert['alert_start'].astimezone(
+                'US/Eastern')
+            alert_end_est = last_alert['alert_end'].astimezone('US/Eastern')
+
+            # Ignore alerts that occur outside of a library's regular hours
+            if (alert_start_est < regular_close_est
+                    and alert_end_est > regular_open_est):
+                # Clamp the closure to the library's regular hours
+                closure_start = max(regular_open_est, alert_start_est)
+                closure_end = min(regular_close_est, alert_end_est)
+
+                # If the stated closure doesn't match what's seen by the
+                # poller, infer the real closure from the polling times.
+                alert_poll_times = set(alert_group['polling_datetime'])
+                expected_alert_poll_times = {
+                    dt for dt in polling_datetimes
+                    if dt > closure_start and dt < closure_end}
+                regular_hours_poll_times = {
+                    dt for dt in polling_datetimes
+                    if dt > regular_open_est and dt < regular_close_est}
+                if not expected_alert_poll_times.issubset(alert_poll_times):
+                    closure_start = min(
+                        alert_poll_times).astimezone('US/Eastern')
+                    closure_end = last_alert['polling_datetime'].astimezone(
+                        'US/Eastern')
+                closure['closure_start'] = closure_start.time().isoformat()
+                closure['closure_end'] = closure_end.time().isoformat()
+                closure['is_full_day'] = regular_hours_poll_times.issubset(
+                    alert_poll_times)
+                closures.append(closure)
+
+    return None if len(closures) == 0 else pd.DataFrame.from_dict(closures)
+
+
+def lambda_handler(event, context):
+    kms_client = KmsClient()
+    redshift_client = RedshiftClient(
+        kms_client.decrypt(os.environ['REDSHIFT_DB_HOST']),
+        os.environ['REDSHIFT_DB_NAME'],
+        kms_client.decrypt(os.environ['REDSHIFT_DB_USER']),
+        kms_client.decrypt(os.environ['REDSHIFT_DB_PASSWORD']))
+    kms_client.close()
+
+    redshift_client.connect()
+    alerts_df = redshift_client.execute_query(build_get_alerts_query())
+    closures = get_closures(alerts_df)
+    try:
+        cursor = redshift_client.conn.cursor()
+        cursor.execute('BEGIN TRANSACTION;')
+        if closures is not None:
+            cursor.write_dataframe(
+                closures, os.environ['REDSHIFT_CLOSURES_TABLE'])
+        cursor.execute('DELETE FROM {};'.format(
+            os.environ['REDSHIFT_CLOSURE_ALERTS_TABLE']))
```

Comment:
Is this configurable because the table name is different in different environments? Any way to protect against deleting all data from the wrong table through misconfiguration? (e.g. establish table name as one of two static options based on some other "environment" param?) If the Redshif user perms only have write access to these two specific tables, I'm less concerned.

Reply:
Updated to hardcode the specific table names. The Redshift user should also only have write access to the specific tables.